# Classification-1（練習用）

2値分類用のサンプルデータであるBreast Cancerを読み込みます。<br>特徴量が全て数値データで欠損値もないキレイなサンプルデータです。

In [1]:
# Load and return the breast cancer wisconsin dataset (classification).
# The breast cancer dataset is a classic and very easy binary classification dataset.
import pandas as pd
from sklearn.datasets import load_breast_cancer

# Set dataframe
dataset = load_breast_cancer()
X = pd.DataFrame(dataset.data, columns=dataset.feature_names)
y = pd.Series(dataset.target, name='y')

# check the shape
print('--------------------------------------------------')
print('X shape: (%i,%i)' %X.shape)
print('--------------------------------------------------')
print(y.value_counts())
print('--------------------------------------------------')
print('y=0 means Marignant(悪性),y=1 means Benign(良性):')
print('--------------------------------------------------')
X.join(y).head()

--------------------------------------------------
X shape: (569,30)
--------------------------------------------------
1    357
0    212
Name: y, dtype: int64
--------------------------------------------------
y=0 means Marignant(悪性),y=1 means Benign(良性):
--------------------------------------------------


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension ...  worst texture  worst perimeter  worst area  \
0                 0.07871 ...          17.33           184.60      2019.0   
1                 0.05667 ...          23.41           158.80      1956.0   
2                 0.05999 ...          25.53           152.50      1709.0   
3                 0.09744 ...          26.50            98.87       567.7   
4                 0.05883 ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  y  
0          0.4601                  0.11890  0  
1          0.2750                  0.08902  0  
2          0.3613                  0.08758  0  
3          0.6638                  0.17300  0  
4          0.2364                  0.07678  0  

[5 rows x 31 columns]

In [2]:
## カラムの表示数を増やすオプション。35列まで表示
pd.set_option('display.max_columns', 35)

In [3]:
X.join(y).head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  radius error  texture error  perimeter error  \
0                 0.07871        1.0950         0.9053            8.589   
1                 0.05667        0.5435         0.7339            3.398   
2                 0.05999        0.7456         0.7869            4.585   
3                 0.09744        0.4956         1.1560            3.445   
4                 0.05883        0.7572         0.7813            5.438   

   area error  smoothness error  compactness error  concavity error  \
0      153.40          0.006399            0.04904          0.05373   
1       74.08          0.005225            0.01308          0.01860   
2       94.03          0.006150            0.04006          0.03832   
3       27.23          0.009110            0.07458          0.05661   
4       94.44          0.011490            0.02461          0.05688   

   concave points error  symmetry error  fractal dimension error  \
0               0.01587         0.03003                 0.006193   
1               0.01340         0.01389                 0.003532   
2               0.02058         0.02250                 0.004571   
3               0.01867         0.05963                 0.009208   
4               0.01885         0.01756                 0.005115   

   worst radius  worst texture  worst perimeter  worst area  worst smoothness  \
0         25.38          17.33           184.60      2019.0            0.1622   
1         24.99          23.41           158.80      1956.0            0.1238   
2         23.57          25.53           152.50      1709.0            0.1444   
3         14.91          26.50            98.87       567.7            0.2098   
4         22.54          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  y  
0                  0.11890  0  
1                  0.08902  0  
2                  0.08758  0  
3                  0.17300  0  
4                  0.07678  0

In [4]:
X.join(y).describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  radius error  texture error  \
count     569.000000              569.000000    569.000000     569.000000   
mean        0.181162                0.062798      0.405172       1.216853   
std         0.027414                0.007060      0.277313       0.551648   
min         0.106000                0.049960      0.111500       0.360200   
25%         0.161900                0.057700      0.232400       0.833900   
50%         0.179200                0.061540      0.324200       1.108000   
75%         0.195700                0.066120      0.478900       1.474000   
max         0.304000                0.097440      2.873000       4.885000   

       perimeter error  area error  smoothness error  compactness error  \
count       569.000000  569.000000        569.000000         569.000000   
mean          2.866059   40.337079          0.007041           0.025478   
std           2.021855   45.491006          0.003003           0.017908   
min           0.757000    6.802000          0.001713           0.002252   
25%           1.606000   17.850000          0.005169           0.013080   
50%           2.287000   24.530000          0.006380           0.020450   
75%           3.357000   45.190000          0.008146           0.032450   
max          21.980000  542.200000          0.031130           0.135400   

       concavity error  concave points error  symmetry error  \
count       569.000000            569.000000      569.000000   
mean          0.031894              0.011796        0.020542   
std           0.030186              0.006170        0.008266   
min           0.000000              0.000000        0.007882   
25%           0.015090              0.007638        0.015160   
50%           0.025890              0.010930        0.018730   
75%           0.042050              0.014710        0.023480   
max           0.396000              0.052790        0.078950   

       fractal dimension error  worst radius  worst texture  worst perimeter  \
count               569.000000    569.000000     569.000000       569.000000   
mean                  0.003795     16.269190      25.677223       107.261213   
std                   0.002646      4.833242       6.146258        33.602542   
min                   0.000895      7.930000      12.020000        50.410000   
25%                   0.002248     13.010000      21.080000        84.110000   
50%                   0.003187     14.970000      25.410000        97.660000   
75%                   0.004558     18.790000      29.720000       125.400000   
max                   0.029840     36.040000      49.540000       251.200000   

        worst area  wor

k近傍法とロジスティック回帰の呼び出し方を学びましょう。KNNはneighborsというクラスから呼び出しています。ロジスティック回帰は重みベクトルと特徴量の一次線形結合をベースとしたアルゴリズムなので、線形モデルから呼び出せます。

In [5]:
# import libraries
from sklearn.neighbors import KNeighborsClassifier                   ## K近傍法（K-Nearest Neighbor）
from sklearn.linear_model import LogisticRegression                  ## ロジスティック回帰（Logistic Regression）正則化含む

回帰でも学びましたが、機械学習のアルゴリズムは特徴量ベクトルを標準化するのが普通です。実装は標準化処理とモデル学習（重みの最適化）のプロセスをひとまとめに記述できるパイプラインが便利なため呼び出します。

In [6]:
# import libraries
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

回帰モデルの評価にはR2スコア等がありました。分類モデルの評価の場合、最もわかりやすいものが正解率（Accuracy）です。正解率は予測対象データ（スコアリングデータ）の全件をNとした時、正例を正しく正例と分類できた数TP、負例を正しく負例と分類できた数TFとすると、 以下の数式で表せます。<br><center>Accuracy = (TP + TF) / N</center><br>つまり、正例・負例を問わず正しく分類されたサンプル数の割合です。sklearnでは正解データ（正解ベクトル）と予測データ（予測ベクトル）を与えると、Accuracyを計算してくれる関数が存在します。それを以下で呼び込みましょう。

In [7]:
# 分類モデルの評価指標計算のための関数の読込
from sklearn.metrics import accuracy_score

以下コードの各行の役割を理解できていることを確認して下さい。

In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

# Holdout
X_train,X_test,y_train,y_test = train_test_split(X,
                                                 y,
                                                 test_size=0.20,
                                                 random_state=1)
# set pipelines for two different algorithms
pipelines ={
    'knn_5': Pipeline([('scl',StandardScaler()),
                       ('est',KNeighborsClassifier())]),
    'knn_50': Pipeline([('scl',StandardScaler()),
                        ('est',KNeighborsClassifier(n_neighbors=50))]),
    'logistic': Pipeline([('scl',StandardScaler()),
                          ('est',LogisticRegression(random_state=1))])    
}

In [9]:
print("X:", X.shape)
print("y:", y.shape)
print("X_train:", X_train.shape)
print("y_train:", y_train.shape)
print("X_test:", X_test.shape)
print("y_test:", y_test.shape)

X: (569, 30)
y: (569,)
X_train: (455, 30)
y_train: (455,)
X_test: (114, 30)
y_test: (114,)


In [10]:
pipelines

{'knn_5': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
            metric_params=None, n_jobs=1, n_neighbors=5, p=2,
            weights='uniform'))]), 'knn_50': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
            metric_params=None, n_jobs=1, n_neighbors=50, p=2,
            weights='uniform'))]), 'logistic': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
           intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
           penalty='l2', random_state=1, solver='liblinear', tol=0.0001,
           verbose=0, warm_start=False))])}

knn_5はデフォルトのK近傍法のパフォーマンス確認、pipe_knn_50は近傍数を増やしたときの性能確認用です。Kを増やすことで予測性能（正解率）がどのように変化するか考えてみましょう。pipe_logisticはロジスティック回帰の性能確認用です。kNNよりは良いだろうというのが通常の期待です。それでは各パイプラインを学習(fit)させましょう。正常終了すれば終了です。

In [11]:
# optimize the parameters of each algorithms
for pipe_name, pipeline in pipelines.items():
    pipeline.fit(X_train,y_train)
    print(pipe_name, ': Fitting Done')

knn_5 : Fitting Done
knn_50 : Fitting Done
logistic : Fitting Done


最後に、３つのパイプラインの分類器としての性能（正解率）を比較します。<br>まずpipelineのpredictの出力ベクトルを見て下さい。<br>y_pred in {0,1}で予測ラベルが付与されていることがわかります。

In [12]:
pipelines['logistic'].predict(X_train)[:10]

array([0, 0, 1, 1, 1, 1, 1, 0, 0, 1])

今回は使いませんが、実務では予測ラベルの他に各クラスに分類される予測確率をよく使います。<b>予測確率を出力させたい場合はpredictをpredic_probaに変更するだけです。</b>最初の列はy=0と分類される予測確率、2番目の列はy=1と分類される予測確率です。

In [14]:
pipelines['logistic'].predict_proba(X_train)[:10]

array([[9.99441634e-01, 5.58366357e-04],
       [9.99962234e-01, 3.77655428e-05],
       [1.12299304e-06, 9.99998877e-01],
       [4.10316174e-04, 9.99589684e-01],
       [1.83025794e-04, 9.99816974e-01],
       [7.88575120e-03, 9.92114249e-01],
       [6.74588243e-02, 9.32541176e-01],
       [9.99956558e-01, 4.34422063e-05],
       [9.99999212e-01, 7.88384042e-07],
       [2.32654465e-02, 9.76734554e-01]])

In [15]:
## X_trainとy_trainと予測値を連結し確認
pred_proba = pd.DataFrame( pipelines['logistic'].predict_proba(X_train) )
pred_proba.columns = ["proba_of_0", "proba_of_1"]

pred = pd.DataFrame( pipelines['logistic'].predict(X_train) )
pred.columns = ["pred"]

train_res_df = X_train.join(y_train).reset_index(drop=True).join(pred_proba).join(pred)  ## joinのためindexをリセットしておく
train_res_df
## 予測値が確率で出力されていることを確認

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         20.66          117.80      991.7          0.10360   
1          20.29         14.34          135.10     1297.0          0.10030   
2           9.00         14.40           56.36      246.3          0.07005   
3          12.21         14.09           78.78      462.0          0.08108   
4          12.34         14.95           78.29      469.1          0.08682   
5          14.86         16.94           94.89      673.7          0.08924   
6          13.54         14.36           87.46      566.3          0.09779   
7          19.89         20.26          130.50     1214.0          0.10370   
8          20.55         20.86          137.80     1308.0          0.10460   
9          11.99         24.89           77.61      441.3          0.10300   
10         12.88         28.92           82.50      514.3          0.08123   
11         23.21         26.97          153.50     1670.0          0.09509   
12         14.87         16.67           98.64      682.5          0.11620   
13         16.84         19.46          108.40      880.2          0.07445   
14         19.59         18.15          130.70     1214.0          0.11200   
15         20.18         23.97          143.70     1245.0          0.12860   
16         12.27         17.92           78.41      466.1          0.08685   
17         15.73         11.28          102.80      747.2          0.10430   
18         12.30         19.02           77.88      464.4          0.08313   
19         19.07         24.81          128.30     1104.0          0.09081   
20         11.08         14.71           70.21      372.7          0.10060   
21         15.75         20.25          102.60      761.3          0.10250   
22         17.54         19.32          115.10      951.6          0.08968   
23         15.71         13.93          102.00      761.7          0.09462   
24         11.22         33.81           70.79      386.8          0.07780   
25         14.99         22.11           97.53      693.7          0.08515   
26         15.05         19.07           97.26      701.9          0.09215   
27         19.40         23.50          129.10     1155.0          0.10270   
28         11.89         21.17           76.39      433.8          0.09773   
29         15.22         30.62          103.40      716.9          0.10480   
..           ...           ...             ...        ...              ...   
425        17.19         22.07          111.60      928.3          0.09726   
426        15.27         12.91           98.17      725.5          0.08182   
427        12.18         14.08           77.25      461.4          0.07734   
428        14.58         13.66           94.29      658.8          0.09832   
429        11.54         10.72           73.73      409.1          0.08597   
430        10.96         17.62           70.79      365.6          0.09687   
431        12.43         17.00           78.60      477.3          0.07557   
432        13.71         20.83           90.20      577.9          0.11890   
433        21.61         22.28          144.40     1407.0          0.11670   
434        16.11         18.05          105.10      813.0          0.09721   
435        14.48         21.46           94.25      648.2          0.09444   
436        11.49         14.59           73.99      404.9          0.10460   
437        23.09         19.83          152.10     1682.0          0.09342   
438        13.86         16.93           90.96      578.9          0.10260   
439        11.06         14.83           70.31      378.2          0.07741   
440        12.25         22.44           78.18      466.5          0.08192   
441        19.73         19.82          130.70     1206.0          0.10620   
442        17.60         23.33          119.00      980.5          0.09289   
443        13.87         16.21           88.52      593.7          0.08743   
444        19.45         19

accuracy_scoreは第一引数に正解ラベル、第二引数に予測ラベル（確率ではない）を指定します。<br>今回はpredictで予測ラベルを与えましょう。

In [16]:
from sklearn.metrics import accuracy_score

scores = {}
for pipe_name, pipeline in pipelines.items():
    scores[(pipe_name,'train')] = accuracy_score(y_train, pipeline.predict(X_train))
    scores[(pipe_name,'test')] = accuracy_score(y_test, pipeline.predict(X_test))

pd.Series(scores).unstack()

test     train
knn_5     0.956140  0.982418
knn_50    0.947368  0.951648
logistic  0.982456  0.991209

k近傍法のkを5から50に増やしましたが、性能が落ちてしまいました。多くのデータを見過ぎて平均値予測に近づき過ぎてしまったようです。予測値が無難過ぎてTrainスコアとTestスコアが近いのも特徴です。ロジスティック回帰は、k近傍法より3pt（0.956 to 0.982）程度、改善することを確認できました。ただし、どのように改善するか（しないか）はデータ次第ですので、代表的なアルゴリズムをいつでも試せる準備をしておくことが、プロジェクト進行において大切です。

In [19]:
scores

{('knn_5', 'train'): 0.9824175824175824,
 ('knn_5', 'test'): 0.956140350877193,
 ('knn_50', 'train'): 0.9516483516483516,
 ('knn_50', 'test'): 0.9473684210526315,
 ('logistic', 'train'): 0.9912087912087912,
 ('logistic', 'test'): 0.9824561403508771}

In [20]:
## 実測値と予測結果をクロス集計
## 予測結果は上で求めたlogistic
## データはtrain
pd.crosstab(train_res_df['y'], train_res_df['pred'])

pred    0    1
y             
0     167    3
1       1  284

In [21]:
## 合計データ数は455。正解率を計算すると
(167 + 284) / 455

## accuracy_score関数で求めたものに一致

0.9912087912087912